src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -25,8 +25,7 @@ namespace Microsoft.AspNetCore.Server.IISIntegration
     internal abstract partial class IISHttpContext : NativeRequestContext, IDisposable
     {
         private const int MinAllocBufferSize = 2048;
-
-        private static bool UpgradeAvailable = (Environment.OSVersion.Version >= new Version(6, 2));
+        private static bool? UpgradeAvailable;
```

Comment:
This will likely need locking or something to make it thread-safe. Since it's static it can be accessed by multiple concurrent requests. You probably don't need locking, just to make sure that it's initialized atomically. It's OK if multiple requests read the server variable, since they should all set the same value.

Reply:
Seem my comment on setting the UpgradeAvailable. I don't think it's needed. All requests that come into the server will return the same value.

Reply:
It's not a bool though, it's a `Nullable<bool>` which has two fields and could be sheared if it's being written while another thread reads it.

Agreed, if you can make it an atomic type, it should be safe without locks, multiple threads may do the same check but they'll all come to the same conclusion. Maybe just use an int with three states: `1` Uninitialized, `2` WebSockets Available, `3` WebSockets Not Available.